File: LeetCode/0640. Solve the Equation/solution.cpp
/**
 * 59 / 59 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 6 MB 
 */
class Solution {
public:
    string solveEquation(string equation) {
        int x = 0, num = 0;
        for (int i = 0, n = equation.size(), sign = 1; i < n; ) {
            if (equation[i] == '=') {
                sign = -1;
                i += 1;
                continue;
            }

            int s = sign;
            if (equation[i] == '+') {
                i += 1;
            } else if (equation[i] == '-') {
                s = -s;
                i += 1;
            }
            int k = 0;
            bool valid = false;
            while (i < n && '0' <= equation[i] && equation[i] <= '9') {
                k = k * 10 + equation[i] - '0';
                valid = true;
                i += 1;
            }
            if (i < n && equation[i] == 'x') {
                x += valid ? s * k : s; 
                i += 1;
            } else {
                num += s * k;
            }
        }
        if (x == 0) {
            if (num == 0) {
                return "Infinite solutions";
            } else {
                return "No solution";
            }
        }
        return "x=" + to_string(-num/x);
    }
};

File: LeetCode/0640. Solve the Equation/solution.py
"""
59 / 59 test cases passed.
Runtime: 52 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def solveEquation(self, equation: str) -> str:
        x = num = 0
        i, n, sign = 0, len(equation), 1
        while i < n:
            if equation[i] == '=':
                sign = -1
                i += 1
                continue

            s = sign
            if equation[i] == '+':
                i += 1
            elif equation[i] == '-':
                s = -s 
                i += 1

            k, valid = 0, False
            while i < n and equation[i].isdigit():
                valid = True
                k = k * 10 + int(equation[i])
                i += 1
            if i < n and equation[i] == 'x':
                x += s * k if valid else s
                i += 1
            else:
                num += s * k
            
        if x == 0:
            return "No solution" if num else "Infinite solutions"
        return f"x={-num//x}"
